File: api/src/test/java/br/com/tradugil/traducao/NormalizadorTest.java
package br.com.tradugil.traducao;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * Estes testes são o contrato que as versões TypeScript e Kotlin do
 * normalizador precisam cumprir. Ao portar a classe, porte os casos junto:
 * uma divergência aqui vira uma gíria que funciona online e some offline.
 */
class NormalizadorTest {

    @Test
    @DisplayName("remove acento e caixa, chegando à mesma chave")
    void normalizaAcentoECaixa() {
        assertThat(Normalizador.normalizar("Ranço")).isEqualTo("ranco");
        assertThat(Normalizador.normalizar("RANÇO")).isEqualTo("ranco");
        assertThat(Normalizador.normalizar("ranco")).isEqualTo("ranco");
        assertThat(Normalizador.normalizar("migué")).isEqualTo("migue");
    }

    @Test
    @DisplayName("descarta pontuação sem colar as palavras vizinhas")
    void descartaPontuacao() {
        assertThat(Normalizador.normalizar("pog!")).isEqualTo("pog");
        assertThat(Normalizador.normalizar("...cringe?")).isEqualTo("cringe");
        assertThat(Normalizador.normalizar("dar,ruim")).isEqualTo("dar ruim");
    }

    @Test
    @DisplayName("preserva o espaço interno dos verbetes de mais de uma palavra")
    void preservaEspacoInterno() {
        assertThat(Normalizador.normalizar("  Dar   Ruim  ")).isEqualTo("dar ruim");
        assertThat(Normalizador.normalizar("No Cap")).isEqualTo("no cap");
    }

    @Test
    @DisplayName("mantém dígitos, porque '67' é verbete")
    void mantemDigitos() {
        assertThat(Normalizador.normalizar("67")).isEqualTo("67");
    }

    @Test
    @DisplayName("trata acento pré-composto e decomposto como o mesmo termo")
    void tratamAcentoComposto() {
        // "ç" chega como um caractere só (U+00E7) ou como "c" mais cedilha
        // combinante (U+0327), dependendo do teclado e do sistema: texto
        // vindo de OCR ou do iOS costuma chegar decomposto. São a mesma
        // palavra para quem digitou, e precisam gerar a mesma chave.
        String preComposto = "ranço";
        String decomposto = "ranço";

        assertThat(preComposto).isNotEqualTo(decomposto);
        assertThat(Normalizador.normalizar(preComposto)).isEqualTo("ranco");
        assertThat(Normalizador.normalizar(decomposto)).isEqualTo("ranco");
    }

    @Test
    void entradaVaziaNaoQuebra() {
        assertThat(Normalizador.normalizar(null)).isEmpty();
        assertThat(Normalizador.normalizar("   ")).isEmpty();
        assertThat(Normalizador.normalizar("!!!")).isEmpty();
    }

    @Test
    @DisplayName("colapsa ênfase até duas letras, não até uma")
    void colapsaEnfase() {
        assertThat(Normalizador.colapsarRepeticoes("mdsss")).isEqualTo("mdss");
        assertThat(Normalizador.colapsarRepeticoes("kkkkkkkk")).isEqualTo("kk");
        assertThat(Normalizador.colapsarRepeticoes("aaaa")).isEqualTo("aa");
    }

    @Test
    @DisplayName("não destrói dígrafo legítimo: 'carro' não vira 'caro'")
    void preservaDigrafo() {
        assertThat(Normalizador.colapsarRepeticoes("carro")).isEqualTo("carro");
        assertThat(Normalizador.colapsarRepeticoes("nossa")).isEqualTo("nossa");
        assertThat(Normalizador.colapsarRepeticoes("kkk")).isEqualTo("kk");
    }
}

File: packages/core-ts/src/normalizador.test.ts
import { describe, expect, it } from 'vitest';
import { colapsarRepeticoes, normalizar } from './normalizador.js';

/**
 * Espelho de `NormalizadorTest.java`. Os casos precisam ser os mesmos, com
 * os mesmos resultados esperados: é assim que se verifica que as duas
 * implementações não divergiram.
 */
describe('normalizar', () => {
  it('remove acento e caixa, chegando à mesma chave', () => {
    expect(normalizar('Ranço')).toBe('ranco');
    expect(normalizar('RANÇO')).toBe('ranco');
    expect(normalizar('ranco')).toBe('ranco');
    expect(normalizar('migué')).toBe('migue');
  });

  it('descarta pontuação sem colar as palavras vizinhas', () => {
    expect(normalizar('pog!')).toBe('pog');
    expect(normalizar('...cringe?')).toBe('cringe');
    expect(normalizar('dar,ruim')).toBe('dar ruim');
  });

  it('preserva o espaço interno dos verbetes de mais de uma palavra', () => {
    expect(normalizar('  Dar   Ruim  ')).toBe('dar ruim');
    expect(normalizar('No Cap')).toBe('no cap');
  });

  it("mantém dígitos, porque '67' é verbete", () => {
    expect(normalizar('67')).toBe('67');
  });

  it('entrada vazia não quebra', () => {
    expect(normalizar(null)).toBe('');
    expect(normalizar(undefined)).toBe('');
    expect(normalizar('   ')).toBe('');
    expect(normalizar('!!!')).toBe('');
  });

  it('trata acento pré-composto e decomposto como o mesmo termo', () => {
    // "ç" pode chegar como um caractere só (U+00E7) ou como "c" + cedilha
    // (U+0063 U+0327), dependendo do teclado e do sistema. O usuário digitou
    // a mesma palavra nos dois casos.
    const preComposto = 'ranço';
    const decomposto = 'ranço';
    expect(preComposto).not.toBe(decomposto);
    expect(normalizar(preComposto)).toBe('ranco');
    expect(normalizar(decomposto)).toBe('ranco');
  });
});

describe('colapsarRepeticoes', () => {
  it('colapsa ênfase até duas letras, não até uma', () => {
    expect(colapsarRepeticoes('mdsss')).toBe('mdss');
    expect(colapsarRepeticoes('kkkkkkkk')).toBe('kk');
    expect(colapsarRepeticoes('aaaa')).toBe('aa');
  });

  it("não destrói dígrafo legítimo: 'carro' não vira 'caro'", () => {
    expect(colapsarRepeticoes('carro')).toBe('carro');
    expect(colapsarRepeticoes('nossa')).toBe('nossa');
    expect(colapsarRepeticoes('kkk')).toBe('kk');
  });

  it('entrada vazia não quebra', () => {
    expect(colapsarRepeticoes(null)).toBe('');
    expect(colapsarRepeticoes('')).toBe('');
  });
});
